src/libmongoc/src/mongoc/mongoc-topology-background-monitoring.c

```diff
@@ -802,6 +802,9 @@ _mongoc_topology_background_monitoring_stop (mongoc_topology_t *topology)
    int i;
    bool is_srv_polling;
 
+   /* Test only assert. Is a noop unless -DMONGOC_ENABLE_TESTING is passed */
+   MONGOC_TEST_ASSERT (COMMON_PREFIX (mutex_is_locked) (&topology->mutex));
```

Comment:
Additionally, mongoc-server-monitor.c has many functions that document "Caller must hold no locks" or "must hold topology lock" that we can apply this assertion to. You may need to rebase on an updated master branch, since that file is relatively recent (from CDRIVER-3535).